packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts

```diff
@@ -518,137 +462,54 @@ export class GutenbergEditorPage {
-			this.page.waitForNavigation( { url: '**/pages/**' } ),
-		] );
-
-		const actions: Promise< unknown >[] = [ navigationPromise ];
-
-		if ( envVariables.VIEWPORT_NAME !== 'mobile' ) {
-			actions.push( frame.click( selectors.desktopDashboardLink ) );
-		} else {
-			actions.push( navbarComponent.clickMySites() );
-		}
-
-		await Promise.all( actions );
+	async exitEditor(): Promise< void > {
+		await this.editorNavSidebarComponent.openSidebar();
+		await this.editorNavSidebarComponent.exitEditor();
 	}
 
 	/* Previews */
 
 	/**
-	 * Click on the `Preview` button on the editor toolbar.
+	 * Launches the Preview.
 	 *
-	 * This method interacts with the mobile implementation of the editor preview,
-	 * which:
-	 * 	1. launch a new tab.
-	 * 	2. load the preview.
+	 * For Desktop viewports, this method will not return any value.
+	 * For Mobile viewports, this method will return a reference to a popup Page.
 	 *
-	 * This method will throw if used in a desktop environment.
-	 *
-	 * @throws {Error} If environment is not 'mobile'.
+	 * @returns {Page|void} Handler for the Page object on mobile. Void otherwise.
 	 */
-	async openPreviewAsMobile(): Promise< Page > {
-		if ( envVariables.VIEWPORT_NAME !== 'mobile' ) {
-			throw new Error( 'This method only works in a mobile environment.' );
-		}
-		const frame = await this.getEditorFrame();
-		const [ popup ] = await Promise.all( [
-			this.page.waitForEvent( 'popup' ),
-			frame.click( selectors.previewButton ),
-		] );
-		await popup.waitForLoadState( 'load' );
-		return popup;
-	}
-
-	/**
-	 * Click on the `Preview` button on the editor toolbar, then select requested the preview option.
-	 *
-	 * This method interacts with the non-mobile implementation of the editor preview,
-	 * which applies an attribute to the editor to simulate the preview environment.
-	 *
-	 * Additionally, this method only works in the desktop browser environment; in a mobile
-	 * environment, the Preview button will launch a new page. For mobile, use
-	 * `GutenbergEditorPage.openPreviewAsMobile` instead.
-	 *
-	 * @param {PreviewOptions} target Preview option to be selected.
-	 * @throws {Error} If environment is 'mobile'.
-	 */
-	async openPreviewAsDesktop( target: PreviewOptions ): Promise< void > {
+	async preview( target?: PreviewOptions ): Promise< Page | void > {
 		if ( envVariables.VIEWPORT_NAME === 'mobile' ) {
-			throw new Error( 'This method only works in a non-mobile environment.' );
+			return await this.editorToolbarComponent.openMobilePreview();
 		}
-		const frame = await this.getEditorFrame();
-		await frame.click( selectors.previewButton );
-		await frame.click( selectors.previewMenuItem( target ) );
-		const handle = await frame.waitForSelector( selectors.previewPane( target ) );
-		await handle.waitForElementState( 'stable' );
+		if ( ! target ) {
+			throw new Error( 'Preview target must be defined.' );
+		}
+
+		await this.editorToolbarComponent.openDesktopPreview( target );
 	}
```

Comment:
Could you share a little more about wanting to collapse the two preview methods across viewports here?

In general, I'm definitely very much in favor of that (and actually just recently put this in the docs!), but I wonder if in this case it makes things cleaner to make an exception and keep them separate. Namely, I'm thinking because those two actions essentially required different method signatures! And collapsing those signatures here means that everyone who calls this method is already going to have to write a bunch of logical handling around the parameter values and differing return types. Given that, I think it might be clearer to just keep the two distinct methods.

What do you think?

Reply:
I wanted to keep things in line with rest of the refactored files where there exists _one_and only one obvious way to perform an action.

Your point about the caller having to write handlers for the `void|Page` return value is true and it just might be that there's no way around this other than splitting the preview methods into two.

Reply:
Fixed in 9d1c1774ecf61e7a2878264695ec9d8fd0c14bf0.